lighthouse-cli/test/smokehouse/frontends/smokehouse-bin.js

```diff
@@ -75,6 +76,47 @@ function getDefinitionsToRun(allTestDefns, requestedIds, {invertMatch}) {
   return smokes;
 }
 
+/**
+ * Parses the cli `shardArg` flag into `shardNumber/shardTotal`. Splits
+ * `testDefns` into `shardTotal` shards and returns the `shardNumber`th shard.
+ * Shards will differ in size by at most 1.
+ * Shard params must be 1≤shardNumber≤shardTotal.
+ * @param {Array<Smokehouse.TestDfn>} testDefns
+ * @param {string=} shardArg
+ * @return {Array<Smokehouse.TestDfn>}
+ */
+function getShardedDefinitions(testDefns, shardArg) {
+  if (!shardArg) return testDefns;
+
+  const errorMessage = `'shard' must be of the form 'n/d' and n and d must be positive integers with 1≤n≤d. Got '${shardArg}'`;
+  const match = /^(?<shardNumber>\d+)\/(?<shardTotal>\d+)$/.exec(shardArg);
+  assert(match && match.groups, errorMessage);
+  const shardNumber = Number(match.groups.shardNumber);
+  const shardTotal = Number(match.groups.shardTotal);
+  assert(shardNumber > 0 && Number.isInteger(shardNumber), errorMessage);
+  assert(shardTotal > 0 && Number.isInteger(shardTotal));
+  assert(shardNumber <= shardTotal, errorMessage);
+
+  // Array is sharded with `Math.ceil(length / shardTotal)` shards first
+  // and then the remaining `Math.floor(length / shardTotal) shards.
+  // e.g. `[0, 1, 2, 3]` split into 3 shards is `[[0, 1], [2], [3]]`.
+  const biggerShardCount = testDefns.length % shardTotal;
+  let shardDefns;
+  if (shardNumber <= biggerShardCount) {
+    const biggerSize = Math.ceil(testDefns.length / shardTotal);
+    const shardEnd = shardNumber * biggerSize;
+    shardDefns = testDefns.slice(shardEnd - biggerSize, shardEnd);
+  } else {
+    const baseSize = Math.floor(testDefns.length / shardTotal);
+    const shardEnd = shardNumber * baseSize + biggerShardCount;
+    shardDefns = testDefns.slice(shardEnd - baseSize, shardEnd);
+  }
```

Comment:
This is clever but hard to reason about :)

What about a simpler approach like: iterate all items, adding to one shard array at a time, and then return the nth shard array? I understand we'd be doing "more" work by constructing all the shards (not important), and that the ordering would be not exactly sequential amongst the shards (maybe important).

Reply:
heh, I was hoping ruby did something in a really simple way for `Array.in_groups` but nah: https://apidock.com/rails/Array/in_groups (expand for Source code)

Reply:
> This is clever but hard to reason about

ha, it even takes care of the ugly `±1`s you get if you do it almost any other way! But fair point.

I don't really want to split up tests by name like they would if distributing them to a shard one at a time (so half of each set would be in each CI smoke test), but what would you think about still doing a full split over all the shards, which would allow asserting that it was constructed correctly (e.g. that `allShards.flat()` is deep equal to the original array)?

Reply:
Yes, keeping this code but with a followup assertion logic that is simple to understand and proves correctness is 💯

Reply:
harder to reason about but has assertions now :)